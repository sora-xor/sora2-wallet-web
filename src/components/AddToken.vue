<template>
  <wallet-base :title="t('addToken.title')" show-back @back="handleBack">
    <div class="add-token">
      <s-tabs :value="currentTab" type="rounded" @change="handleChangeTab">
        <s-tab
          v-for="tab in AddTokenTabs"
          :key="tab"
          :label="t(`addToken.${tab}.title`)"
          :name="tab"
        />
      </s-tabs>
      <component :is="currentTab" @add-token="handleOpenNotification" />
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Action } from 'vuex-class'

import TranslationMixin from './mixins/TranslationMixin'
import WalletBase from './WalletBase.vue'
import AddTokenSearch from './AddTokenSearch.vue'
import AddTokenCustom from './AddTokenCustom.vue'
import { RouteNames, AddTokenTabs } from '../consts'

@Component({
  components: {
    WalletBase,
    AddTokenSearch,
    AddTokenCustom
  }
})
export default class AddToken extends Mixins(TranslationMixin) {
  readonly AddTokenTabs = AddTokenTabs

  @Action navigate

  currentTab = AddTokenTabs.Search

  handleChangeTab (value: AddTokenTabs): void {
    this.currentTab = value
  }

  handleBack (): void {
    this.navigate({ name: RouteNames.Wallet })
  }

  handleOpenNotification (): void {
    this.$notify({
      message: this.t('addToken.success'),
      title: this.t('successText'),
      type: 'success'
    })
  }
}
</script>

<style lang="scss">
.add-token {
  @include custom-tabs;
}
</style>
