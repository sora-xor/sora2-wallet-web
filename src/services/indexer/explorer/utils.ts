import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
