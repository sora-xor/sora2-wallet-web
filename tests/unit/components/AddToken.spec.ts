import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import AddToken from '@/components/AddToken.vue'
import { TranslationMock, SoramitsuElementsImport } from '../../utils'

const localVue = createLocalVue()
localVue.use(Vuex)
const store = new Vuex.Store({
  modules: {
    Router: {
      actions: {
        navigate: jest.fn()
      }
    }
  } as any
})

describe('AddToken.vue', () => {
  beforeEach(() => {
    SoramitsuElementsImport(localVue)
    TranslationMock(AddToken)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(AddToken, { localVue, store })
    expect(wrapper.element).toMatchSnapshot()
  })
})
