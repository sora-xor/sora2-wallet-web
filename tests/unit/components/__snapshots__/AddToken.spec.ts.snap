// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToken.vue should renders correctly 1`] = `
<wallet-base-stub
  showback="true"
  title=""
>
  <div
    class="add-token"
  >
    <s-tabs-stub
      position="top"
      type="rounded"
      value="AddTokenSearch"
    >
      <s-tab-stub
        name="AddTokenSearch"
      />
      <s-tab-stub
        name="AddTokenCustom"
      />
    </s-tabs-stub>
     
    <addtokensearch-stub />
  </div>
</wallet-base-stub>
`;
